File: tooltipview/src/main/java/com/cunoraz/tooltipview/TooltipView.java
package com.cunoraz.tooltipview;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.support.annotation.*;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class TooltipView extends LinearLayout {

    private static final int NOT_PRESENT = Integer.MIN_VALUE;
    private String toolTipTitle;
    private String toolTipMessage;
    private int arrowHeight;
    private int arrowWidth;
    private int cornerRadius;
    private int anchoredViewId;
    private int tooltipBgColor;
    private ArrowLocation arrowLocation;
    private int arrowPositioning;
    private Paint paint;
    private Path tooltipPath;

    private TextView titleTextView;
    private TextView messageTextView;

    public TooltipView(Context context) {
        super(context);
        init(null, 0);
    }

    public TooltipView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public TooltipView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        setWillNotDraw(false);
        Resources res = getResources();
        TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.TooltipView, defStyle, 0);
        try {
            toolTipTitle = a.getString(R.styleable.TooltipView_toolTipTitle);
            toolTipMessage = a.getString(R.styleable.TooltipView_toolTipMessage);
            anchoredViewId = a.getResourceId(R.styleable.TooltipView_anchoredView, View.NO_ID);
            tooltipBgColor = a.getColor(R.styleable.TooltipView_tooltipBgColor, res.getColor(R.color.default_tooltip_bg_color));
            cornerRadius = getDimension(a, R.styleable.TooltipView_cornerRadius, R.dimen.tooltip_default_corner_radius);
            arrowHeight = getDimension(a, R.styleable.TooltipView_arrowHeight, R.dimen.tooltip_default_arrow_height);
            arrowWidth = getDimension(a, R.styleable.TooltipView_arrowWidth, R.dimen.tooltip_default_arrow_width);
            arrowPositioning = a.getInteger(R.styleable.TooltipView_arrowLocation, res.getInteger(R.integer.tooltip_default_arrow_location));
            arrowLocation = arrowPositioning == Location.TOP ? new TopArrowLocation() : new BottomArrowLocation();

        } finally {
            a.recycle();

            LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            if (inflater != null) {
                inflater.inflate(R.layout.tooltip_with_title, this, true);
                initViews();
                setTexts(toolTipTitle, toolTipMessage);
            }
        }
    }

    private void initViews() {
        titleTextView = findViewById(R.id.titleTextView);
        messageTextView = findViewById(R.id.messageTextView);
    }

    private void setTexts(String toolTipTitle, String toolTipMessage) {
        titleTextView.setText(toolTipTitle);
        messageTextView.setText(toolTipMessage);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (arrowPositioning == Location.TOP) {
            setPadding(0, arrowHeight, 0, 0);
        } else {
            setPadding(0, 0, 0, arrowHeight);
        }
        Log.d("onAttachedToWindow", "onAttachedToWindow: ");
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        setMeasuredDimension(getMeasuredWidth(), getMeasuredHeight() + arrowHeight);
    }

    @Override
    public void invalidate() {
        super.invalidate();
        tooltipPath = null;
        paint = null;
    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        if (tooltipPath == null || paint == null) {
            arrowLocation.configureDraw(this, canvas);
        }
        canvas.drawPath(tooltipPath, paint);
        super.onDraw(canvas);
    }

    Paint getTooltipPaint() {
        return paint;
    }

    void setPaint(Paint paint) {
        this.paint = paint;
    }

    Path getTooltipPath() {
        return tooltipPath;
    }

    void setTooltipPath(Path tooltipPath) {
        this.tooltipPath = tooltipPath;
    }

    public int getArrowHeight() {
        return arrowHeight;
    }

    public void setArrowHeight(int arrowHeight) {
        this.arrowHeight = arrowHeight;
        invalidate();
    }

    public void setArrowHeightResource(@DimenRes int resId) {
        arrowHeight = getResources().getDimensionPixelSize(resId);
        invalidate();
    }

    public int getArrowWidth() {
        return arrowWidth;
    }

    public void setArrowWidth(int arrowWidth) {
        this.arrowWidth = arrowWidth;
        invalidate();
    }

    public void setArrowWidthResource(@DimenRes int resId) {
        arrowWidth = getResources().getDimensionPixelSize(resId);
        invalidate();
    }

    public int getCornerRadius() {
        return cornerRadius;
    }

    public void setCornerRadius(int cornerRadius) {
        this.cornerRadius = cornerRadius;
        invalidate();
    }

    public void setCornerRadiusResource(@DimenRes int resId) {
        cornerRadius = getResources().getDimensionPixelSize(resId);
        invalidate();
    }

    public int getAnchoredViewId() {
        return anchoredViewId;
    }

    public void setAnchoredViewId(@IdRes int anchoredViewId) {
        this.anchoredViewId = anchoredViewId;
        invalidate();
    }

    public int getTooltipBgColor() {
        return tooltipBgColor;
    }

    public void setTooltipBgColor(int tooltipBgColor) {
        this.tooltipBgColor = tooltipBgColor;
        invalidate();
    }

    public void setArrowPositioning(int arrowPositioning) {
        this.arrowPositioning = arrowPositioning;
        invalidate();
    }

    private int getDimension(TypedArray typedArray,
                             @StyleableRes int styleableId,
                             @DimenRes int defaultDimension) {
        int result = typedArray.getDimensionPixelSize(styleableId, NOT_PRESENT);
        if (result == NOT_PRESENT) {
            result = getResources().getDimensionPixelSize(defaultDimension);
        }
        return result;
    }

    @IntDef({Location.TOP, Location.BOTTOM})
    @Retention(RetentionPolicy.SOURCE)
    @interface Location {
        int TOP = 0;
        int BOTTOM = 1;
    }
}

File: tooltipview/src/main/java/com/cunoraz/tooltipview/TooltipView.kt
package com.cunoraz.tooltipview

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.support.annotation.*
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView


class TooltipView : LinearLayout {
    private var toolTipTitle: String? = null
    private var toolTipMessage: String? = null

    private lateinit var arrowLocation: ArrowLocation
    private lateinit var titleTextView: TextView
    private lateinit var messageTextView: TextView
    private lateinit var closeButton: ImageView

    private var arrowPosition: Int = 0
    private var closeButtonGravity: Int = 0
    private var closeButtonVisibility: Int = 0
    internal var arrowHeight: Int = 0
    internal var arrowWidth: Int = 0
    internal var cornerRadius: Int = 0
    internal var anchoredViewId: Int = 0
    internal var tooltipBgColor: Int = 0
    internal var tooltipPaint: Paint? = null
    internal var tooltipPath: Path? = null


    constructor(context: Context) : super(context) {
        init(null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        setWillNotDraw(false)
        val res = resources
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.TooltipView, defStyle, 0)
        try {
            with(typedArray) {
                toolTipTitle = getString(R.styleable.TooltipView_toolTipTitle)
                toolTipMessage = getString(R.styleable.TooltipView_toolTipMessage)
                anchoredViewId = getResourceId(R.styleable.TooltipView_anchoredView, View.NO_ID)
                tooltipBgColor = getColor(
                    R.styleable.TooltipView_tooltipBgColor,
                    ContextCompat.getColor(context, R.color.default_tooltip_bg_color)
                )
                cornerRadius = getDimension(
                    typedArray,
                    R.styleable.TooltipView_cornerRadius,
                    R.dimen.tooltip_default_corner_radius
                )
                arrowHeight = getDimension(
                    typedArray,
                    R.styleable.TooltipView_arrowHeight,
                    R.dimen.tooltip_default_arrow_height
                )
                arrowWidth = getDimension(
                    typedArray,
                    R.styleable.TooltipView_arrowWidth,
                    R.dimen.tooltip_default_arrow_width
                )
                arrowPosition = getInteger(
                    R.styleable.TooltipView_arrowLocation,
                    res.getInteger(R.integer.tooltip_default_arrow_location)
                )
                arrowLocation = if (arrowPosition == LOCATION_TOP) TopArrowLocation() else BottomArrowLocation()
                closeButtonGravity = getInteger(
                    R.styleable.TooltipView_closeButtonGravity,
                    res.getInteger(R.integer.tooltip_default_close_button_gravity)
                )
                closeButtonVisibility = getInteger(R.styleable.TooltipView_closeButtonVisibility, VISIBILITY_VISIBLE)
            }
        } finally {
            typedArray.recycle()

            val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            inflater.inflate(R.layout.tooltip_with_title, this, true)

            initViews()

        }
    }

    private fun initViews() {
        titleTextView = findViewById(R.id.titleTextView)
        messageTextView = findViewById(R.id.messageTextView)
        closeButton = findViewById(R.id.closeButton)


        setCloseButtonProperties()
        setTexts()
    }

    private fun setCloseButtonProperties() {
        if (closeButtonVisibility == VISIBILITY_GONE) {
            closeButton.visibility = View.GONE
            return
        }

        val layoutParams = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.WRAP_CONTENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        )
        layoutParams.gravity = when (closeButtonGravity) {

            GRAVITY_TOP -> Gravity.TOP

            GRAVITY_CENTER -> Gravity.CENTER_VERTICAL

            GRAVITY_BOTTOM -> Gravity.BOTTOM

            else -> Gravity.TOP
        }

        closeButton.layoutParams = layoutParams
    }

    private fun setTexts() {
        if (toolTipTitle == null) {
            titleTextView.visibility = View.GONE
        } else {
            titleTextView.text = toolTipTitle
            titleTextView.visibility = View.VISIBLE
        }

        if (toolTipMessage == null) {
            messageTextView.visibility = View.GONE
        } else {
            messageTextView.text = toolTipMessage
            messageTextView.visibility = View.VISIBLE
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (arrowPosition == LOCATION_TOP) {
            setPadding(0, arrowHeight, 0, 0)
        } else {
            setPadding(0, 0, 0, arrowHeight)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        setMeasuredDimension(measuredWidth, measuredHeight + arrowHeight)
    }

    override fun invalidate() {
        super.invalidate()
        tooltipPath = null
        tooltipPaint = null
    }

    override fun onDraw(canvas: Canvas) {
        if (tooltipPath == null || tooltipPaint == null) {
            arrowLocation.configureDraw(this, canvas)
        }
        canvas.drawPath(tooltipPath!!, tooltipPaint!!)
        super.onDraw(canvas)
    }

    internal fun setPaint(paint: Paint) {
        this.tooltipPaint = paint
    }

    fun getArrowHeight(): Int {
        return arrowHeight
    }

    fun setArrowHeight(arrowHeight: Int) {
        this.arrowHeight = arrowHeight
        invalidate()
    }

    fun setArrowHeightResource(@DimenRes resId: Int) {
        arrowHeight = resources.getDimensionPixelSize(resId)
        invalidate()
    }

    fun getArrowWidth(): Int {
        return arrowWidth
    }

    fun setArrowWidth(arrowWidth: Int) {
        this.arrowWidth = arrowWidth
        invalidate()
    }

    fun setArrowWidthResource(@DimenRes resId: Int) {
        arrowWidth = resources.getDimensionPixelSize(resId)
        invalidate()
    }

    fun getCornerRadius(): Int {
        return cornerRadius
    }

    fun setCornerRadius(cornerRadius: Int) {
        this.cornerRadius = cornerRadius
        invalidate()
    }

    fun setCornerRadiusResource(@DimenRes resId: Int) {
        cornerRadius = resources.getDimensionPixelSize(resId)
        invalidate()
    }

    fun getAnchoredViewId(): Int {
        return anchoredViewId
    }

    fun setAnchoredViewId(@IdRes anchoredViewId: Int) {
        this.anchoredViewId = anchoredViewId
        invalidate()
    }

    fun getTooltipBgColor(): Int {
        return tooltipBgColor
    }

    fun setTooltipBgColor(tooltipBgColor: Int) {
        this.tooltipBgColor = tooltipBgColor
        invalidate()
    }

    fun setArrowPosition(arrowPosition: Int) {
        arrowLocation = if (arrowPosition == LOCATION_TOP) TopArrowLocation() else BottomArrowLocation()
        if (arrowPosition == LOCATION_TOP) { // magic number will be dynamic
            (findViewById<LinearLayout>(R.id.rootLayout)).setPadding(32, arrowHeight + 32, 32, 0)
        } else {
            (findViewById<LinearLayout>(R.id.rootLayout)).setPadding(32, 32, 32, arrowHeight)
        }
        invalidate()
    }

    fun setTitle(title: String) {
        titleTextView.text = title
        titleTextView.visibility = View.VISIBLE
    }

    fun getTitle() = toolTipTitle

    fun setMessage(message: String) {
        messageTextView.text = message
        messageTextView.visibility = View.VISIBLE
    }

    fun getMessage() = toolTipMessage

    private fun getDimension(
        typedArray: TypedArray,
        @StyleableRes styleableId: Int,
        @DimenRes defaultDimension: Int
    ): Int {
        var result = typedArray.getDimensionPixelSize(styleableId, NOT_PRESENT)
        if (result == NOT_PRESENT) {
            result = resources.getDimensionPixelSize(defaultDimension)
        }
        return result
    }


    companion object {

        const val VISIBILITY_VISIBLE = 0
        const val VISIBILITY_GONE = 1

        const val LOCATION_TOP = 0
        const val LOCATION_BOTTOM = 1

        const val GRAVITY_TOP = 0
        const val GRAVITY_CENTER = 1
        const val GRAVITY_BOTTOM = 2

        private const val NOT_PRESENT = Integer.MIN_VALUE
    }
}
